lib/core/utils/qsa.js

```diff
@@ -194,27 +173,80 @@ convertExpressions = function (expressions) {
 	});
 };
 
-matchExpressions = function  (domTree, expressions, recurse) {
-	return expressions.reduce((collected, exprArr) => {
-		var candidates = domTree;
-		exprArr.forEach((exp, index) => {
-			recurse = exp.combinator === '>' ? false : recurse;
-			if ([' ', '>'].includes(exp.combinator) === false) {
-				throw new Error('axe.utils.querySelectorAll does not support the combinator: ' + exp.combinator);
+function createLocalVariables (nodes, anyLevel, thisLevel, parentShadowId) {
+	let retVal = {
+		nodes: nodes.slice(),
+		anyLevel: anyLevel,
+		thisLevel: thisLevel,
+		parentShadowId: parentShadowId
+	};
+	retVal.nodes.reverse();
+	return retVal;
+}
+
+function matchesSelector (node, exp) {
+	return (matchesTag(node.actualNode, exp[0]) &&
+			matchesClasses(node.actualNode, exp[0]) &&
+			matchesAttributes(node.actualNode, exp[0]) &&
+			matchesId(node.actualNode, exp[0]) &&
+			matchesPseudos(node, exp[0])
+		);
+}
+
+matchExpressions = function  (domTree, expressions, recurse, parentShadowId, filter) {
```

Comment:
Why did you need to pass the `parentShadowId` in separately? Looks to me like you could've just take that from `domTree`.